substrate/frame/opf/src/lib.rs

```diff
@@ -0,0 +1,669 @@
+
+		/// Project removed from whitelisted projects list
+		ProjectUnlisted { project_id: ProjectId<T> },
+
+		/// Project Funding Accepted by voters
+		ProjectFundingAccepted { project_id: ProjectId<T>, amount: BalanceOf<T> },
+
+		/// Reward claim has expired
+		ExpiredClaim { expired_when: ProvidedBlockNumberFor<T>, project_id: ProjectId<T> },
+
+		/// Project Funding rejected by voters
+		ProjectFundingRejected { project_id: ProjectId<T> },
+
+		/// A new voting round started
+		VotingRoundStarted { round_number: u32 },
+
+		/// The users voting period ended. Reward calculation will start.
+		VoteActionLocked { round_number: u32 },
+
+		/// The voting round ended
+		VotingRoundEnded { round_number: u32 },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Not enough Funds in the Pot
+		InsufficientPotReserves,
+		/// The funds transfer operation failed
+		TransferFailed,
+		/// Spend or Spend index does not exists
+		InexistentSpend,
+		/// No project found under this project_id
+		NoProjectAvailable,
+		/// The Funds transfer failed
+		FailedSpendOperation,
+		/// Still not in claiming period
+		NotClaimingPeriod,
+		/// Still not in funds unlock period
+		NotUnlockPeriod,
+		/// An invalid result  was returned
+		InvalidResult,
+		/// The reward calculation failed due to an internal error
+		FailedRewardCalculation,
+		/// Voting round is over
+		VotingRoundOver,
+		/// This voting round does not exists
+		NoRoundFound,
+		/// Maximum number of projects submission for reward distribution as been reached
+		MaximumProjectsNumber,
+		/// Another project has already been submitted under the same project_id
+		SubmittedProjectId,
+		/// Project batch already submitted
+		BatchAlreadySubmitted,
+		/// Requested vote data do not exist
+		NoVoteData,
+		/// Not enough funds to process the transaction
+		NotEnoughFunds,
+		/// This referendum does not exists
+		ReferendumNotFound,
+		/// The dispatch call failed
+		FailedToDispatchCall,
+		/// The reward has already been claimed
+		AlreadyClaimed,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<SystemBlockNumberFor<T>> for Pallet<T> {
+		fn on_idle(_n: SystemBlockNumberFor<T>, remaining_weight: Weight) -> Weight {
+			Self::on_idle_function(remaining_weight)
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// OPF Projects registration
+		///
+		/// ## Dispatch Origin
+		///
+		/// Must be AdminOrigin
+		///
+		/// ## Details
+		///
+		/// From this extrinsic only AdminOrigin can register whitelisted projects.
+		///
+		/// ### Parameters
+		/// - `projects_id`: The accounts that might be funded.
+		///
+		/// ### Errors
+		/// - [`Error::<T>::MaximumProjectsNumber`]: Maximum number of project subscriptions reached
+		///  
+		/// ## Events
+		/// Emits [`Event::<T>::Projectslisted`].
+		#[pallet::call_index(1)]
+		#[pallet::weight(<T as Config>::WeightInfo::register_projects_batch(T::MaxProjects::get()))]
+		pub fn register_projects_batch(
```

Comment:
`register_projects_batch` shouldn't start a round, projects are registered independently from rounds, OpenGov will register project at any time, synchronizing it with rounds will make it difficult.

AFAIK in the spec what I read was that rounds are going, and OpenGov can register some new projects, if no new projects is coming, rounds should still operate, and OpenGov will probably register project actually one by one, so it is better not to wait for a new rounds everytime. (Because OpenGov wants to vote on project one by one most of the time)

Reply:
If I understand correctly, this would mean that every time a project is submitted in a round and voted on, reward calculation as to be done again and all whitelisted project rewards/spendInfo have to be updated.

Reply:
no, if a project is whitelisted, maybe we can start the poll for this new project, or wait for the next round, both solution are good to me.
Rewards are calculated for a round at the end of the round and are never modified after being calculated.

The flow is:
* round ends, polls for current whitelisted project are ended, reward is calculated given the result of those polls, new polls are created for all whitelisted project, votes from previous polls are resubmitted to the new polls.
* round is on-going, OpenGov can whitelist new projects, or remove some projects from the whitelist.
* repeat